### PR TITLE
Add gcp constraint framework target

### PR DESCRIPTION
**Pull request by briantkennedy**
This creates a constriant framework target for FCV to use for integrating with the opa constraint framework.

### CODE REVIEW

**Comment by briantkennedy** [comment]
@joecheuk or @morgante will either of you have time to take a look at this?

**Comment by morgante** [comment]
I can look too, but wondering if @maxsmythe could provide a review given he's more familiar with the official CF requirements?

File: pkg/gcptarget/gcptarget.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+					},
+				},
+			},
+		},
+	}
+}
+
+// GetName implements client.TargetHandler
+func (g *GCPTarget) GetName() string {
+	// Done
```
Comment (maxsmythe) [inline pkg/gcptarget/gcptarget.go]: dangling comment?

File: pkg/gcptarget/gcptarget.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	return Name
+}
+
+// Library implements client.TargetHandler
+func (g *GCPTarget) Library() *template.Template {
+	// TODO - lib impl
```
Comment (maxsmythe) [inline pkg/gcptarget/gcptarget.go]: TODO == done?

File: pkg/gcptarget/gcptarget.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	return false, "", nil, errors.Errorf("Storing data for referential constraint eval is not supported at this time.")
+}
+
+// HandleReview implements client.TargetHandler
+func (g *GCPTarget) HandleReview(obj interface{}) (bool, interface{}, error) {
+	switch asset := obj.(type) {
+	case *validator.Asset:
+		return g.handleAsset(asset)
+	case map[string]interface{}:
+		return true, asset, nil
```
Comment (maxsmythe) [inline pkg/gcptarget/gcptarget.go]: Is there something (field, value, etc.) we can check to make sure it's the correct map[string]interface{} (e.g., not just some random unmarshalled JSON)?

File: pkg/gcptarget/gcptarget.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	err := json.Unmarshal(buf.Bytes(), &f)
+	if err != nil {
+		return false, nil, errors.Wrapf(err, "marshalling from json with asset %s: %v", asset.Name, asset)
+	}
+	return true, f, nil
+}
+
+// HandleViolation implements client.TargetHandler
+func (g *GCPTarget) HandleViolation(result *types.Result) error {
+	// Done? - this mutates the result after it's provided, we may have to add metadata
```
Comment (maxsmythe) [inline pkg/gcptarget/gcptarget.go]: You may want to populate result.Resource{}, though this is not strictly necessary if you always know the input resource.

File: pkg/gcptarget/gcptarget.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	// to the result
+	return nil
+}
+
+// ValidateConstraint implements client.TargetHandler
+func (g *GCPTarget) ValidateConstraint(constraint *unstructured.Unstructured) error {
+	// TODO: figure out what semantic validation we need here, maybe validate the match criteria
```
Comment (maxsmythe) [inline pkg/gcptarget/gcptarget.go]: The only thing I can think of here is to make sure the strings are well formed (e.g., non-`*` characters followed by a `*`)

File: pkg/gcptarget/library.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	spec := get_default(constraint, "spec", {})
+	match := get_default(spec, "match", {})
+
+	# Default matcher behavior is to match everything.
+	target := get_default(match, "target", ["organization/*"])
+	# TODO: Retire the "gcp" wrapper.
+	# See https://github.com/forseti-security/config-validator/issues/42
+	gcp := get_default(match, "gcp", {})
+	gcp_target := get_default(gcp, "target", target)
+	re_match(gcp_target[_], asset.ancestry_path)
```
Comment (maxsmythe) [inline pkg/gcptarget/library.go]: Are we meant to be using regex matching or prefix and/or glob matching?

File: pkg/gcptarget/library.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+# they are explicitly tested for
+has_field(object, field) = true {
+  object[field] == false
+}
+has_field(object, field) = false {
+  not object[field]
+  not object[field] == false
+}
+
+`
```
Comment (maxsmythe) [inline pkg/gcptarget/library.go]: any Rego tests for this source code?

File: pkg/gcptarget/gcptarget.go (hunk 1 above)
Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: fixed

File: pkg/gcptarget/gcptarget.go (hunk 2 above)
Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: done

File: pkg/gcptarget/gcptarget.go (hunk 3 above)
Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: done

File: pkg/gcptarget/gcptarget.go (hunk 4 above)
Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: We're always going to be passing the input for now.  I've added a TODO for doing this once Audit gets implemented.

Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: This is going to be identical to result.Review, now populated.

File: pkg/gcptarget/gcptarget.go (hunk 5 above)
Comment (briantkennedy) [reply pkg/gcptarget/gcptarget.go]: added regex compilation to this

File: pkg/gcptarget/library.go (hunk 6 above)
Comment (briantkennedy) [reply pkg/gcptarget/library.go]: turns out it's using regex, but the docs indicate glob, it's going to be interesting.

File: pkg/gcptarget/library.go (hunk 7 above)
Comment (briantkennedy) [reply pkg/gcptarget/library.go]: just being tested via the unit test stuff

Comment (maxsmythe) [reply pkg/gcptarget/library.go]: ack

File: pkg/gcptarget/library.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+matching_constraints[constraint] {
+	asset := input.review
+	constraint := {{.ConstraintsRoot}}[_][_]
+	spec := get_default(constraint, "spec", {})
+	match := get_default(spec, "match", {})
+
+	# Default matcher behavior is to match everything.
+	target := get_default(match, "target", ["organizations/.*"])
+  re_match(target[_], asset.ancestry_path)
```
Comment (maxsmythe) [inline pkg/gcptarget/library.go]: nit: indentation

File: pkg/gcptarget/library.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	review := {"msg": "unsupported operation"}
+	constraint := {
+		"msg": "unsupported operation",
+		"kind": "invalid",
+	}
+}
+
+autoreject_review[rejection] {
+	false
+  rejection := {
```
Comment (maxsmythe) [inline pkg/gcptarget/library.go]: nit: indentation

File: pkg/gcptarget/library.go (hunk 6 above)
Comment (maxsmythe) [reply pkg/gcptarget/library.go]: Rego also has Glob:

https://www.openpolicyagent.org/docs/latest/policy-reference/#glob

And the prefix() check is an option for simplicity.

I don't think it's my place to say any one of these is better than the other, however.

File: pkg/gcptarget/library.go (hunk 7 above)
Comment (maxsmythe) [reply pkg/gcptarget/library.go]: ack

File: pkg/gcptarget/gcptarget.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+		}
+		_, foundIam, err := unstructured.NestedMap(asset, "iam_policy")
+		if err != nil {
+			return false, nil, err
+		}
+		if !foundIam && !foundResource {
+			return false, nil, nil
+		}
+		if foundIam && foundResource {
+			return false, nil, errors.Errorf("malformed asset has iam_policy and resource: %v", asset)
```
Comment (morgante) [inline pkg/gcptarget/gcptarget.go]: I'm not sure if this actually an error? I don't see any intrinsic issue with one asset having both.

**Pull request merged**